### DISCUSSION TITLE
Why does my NextAuthJS discord login work in the test environment but cannot be built to production? (NextJS 13)

### DISCUSSION THREAD

**Discussion by devvf** [vercel/next.js#50511] [Help] [answered]
### Summary

I am trying to deploy a NextJS application to vercel. This is a test sign in model using AuthJS. When I run the program using `npm run dev`it works fine and I am able to log in to my discord account which is connected to an application on the developer portal. However, when I deploy it to Vercel and try and build the application for production I receive this error. 
```
Type error: Route "app/api/auth/[...nextauth]/route.ts" does not match the required types of a Next.js Route.
  "authOptions" is not a valid Route export field.
```
For context I am using NextJS 13 and it's app folder method of rendering pages along with the NextAuth.js.

```
import NextAuth from "next-auth";
import type { NextAuthOptions } from "next-auth";
import DiscordProvider from "next-auth/providers/discord";

export const authOptions: NextAuthOptions = {
  providers: [
    DiscordProvider({
      clientId: process.env.DISCORD_CLIENT_ID!,
      clientSecret: process.env.DISCORD_CLIENT_SECRET!,
    })
  ]
};

const handler = NextAuth(authOptions);
export { handler as GET, handler as POST }
```
This is my code found inside the route, and the relative path to this folder is `app/api/auth/[...nextauth]/route.ts` which doesn't seem to have any problems when I run the program, only when trying to build it. I am not sure what to do as I am new to web development and have not encountered such an issue before and am perplexed. Any help is vastly appreciated.

**Comment by devjiwonchoi** [comment]
When you are in the development stage of your Next.js project, Typescript is allowed and is recognized during the runtime.
>In the development stage, Next.js optimizes for the developer and their experience building the application. It comes with features that aim to improve the Developer Experience such the [TypeScript](https://nextjs.org/docs/basic-features/typescript) and [ESLint integration](https://nextjs.org/docs/basic-features/eslint), [Fast Refresh](https://nextjs.org/docs/basic-features/fast-refresh), and more.

On the other hand, when you `next build` your project, Typescript is also compiled into Javascript in your project and is stored in the `build` directory (`.next` as default).

In your case, your `route.ts` has been compiled to `route.js`, but the path that is stated in `NextAuth.js` is `route.ts`.
> Type error: Route "app/api/auth/[...nextauth]/route.ts" does not match the required types of a Next.js Route.

When you are working with Typescript, you do not need to write the filename extension such as `.ts`.

In order to resolve your issue, try to remove the file extension `.ts` of your route path into 
`app/api/auth/[...nextauth]/route`.

**Reply by devvf** [reply]
Hello! Thank you for your help. When I make that change I am now able to deploy the app to production, however my button now merely returns the link `https://spacecases-remake.vercel.app/api/auth/error` rather than the sign in page as it did on the test environment. In addition I can no longer run the code in the test environment as it says 

```
Module parse failed: Unexpected token (2:12)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
| import NextAuth from "next-auth";
> import type { NextAuthOptions } from "next-auth";
| import DiscordProvider from "next-auth/providers/discord";
| 

Import trace for requested module:
./app/api/auth/[...nextauth]/route
```

I am not sure what to do as this seems to contradict what you said about the file extension not mattering - any further help is greatly appreciated thank you.

**Reply by devjiwonchoi** [reply]
Seems like you are using `webpack.js` and if so, you need to configure your `webpack.config.js` for Typescript.
```
npm install --save-dev typescript ts-loader

// or

yarn add -D typescript ts-loader
```
```
// webpack.config.js

module.exports = {
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        use: 'ts-loader',
        exclude: /node_modules/,
      },
    ],
  },
  resolve: {
    extensions: ['.tsx', '.ts', '.js'],
  },
};
```

There might be some other options you need to add for `webpack.js`, but for now these are the essential configuration for your current situation.

**Reply by devvf** [reply]
Thank you for the help. Unfortuntely, however the same issue still persists, I'm not sure if it's my fault I might be doing something wrong - if it is possible for you would you be able to look at the source code here in the repository? https://github.com/goat6/spacecases-remake
Also the current deployed version with your recommended changes can be found here https://spacecases-remake.vercel.app/
Thank you very much for your time.

**Reply by devjiwonchoi** [reply]
Since I cannot refactor every single file of your project, I have modified essential files that are causing build errors.

First of all, your `app/api/auth/[...nextauth]/route.ts` was the problem since in Next.js, the file name `route` is reserved as a convention for `API Routes`. 

This is why it caused an error  `Type error: Route "app/api/auth/[...nextauth]/route.ts" does not match the required types of a Next.js Route.`

I have renamed it to `auth.ts` and it is building fine.

For more information, see [here](https://nextjs.org/docs/app/api-reference/file-conventions/route).

Second, you don't need to add the `"app/api/auth/[...nextauth]/route"` into tsconfig.json since it is covered in `"**/*.ts"`.

Third, your `next.config.js` was invalid next.config format such as 
```
images: {
    domains: ['cdn.discordapp.com'],
  },
```
, and it appears to be unnecessary at this point. If you want to add the correct configuration to your `next.config`, [see here](https://nextjs.org/docs/app/api-reference/next-config-js).

Finally, your `node_modules` had unnecessary dependencies for production in which `@types` files and `eslint` files could be moved to the `devDependencies` in order to reduce your project build size.

I've found some typos such as `favicon.io` but I will leave the rest up to your choice.

One last thing for advice is that do not include `.env` files into the `.git` repository, since the `environment variables` are meant to be kept secret since other "attackers" might be able to use your essential API keys and so forth.

In summary,
1. Don't use `route.ts` unless you are intending to use the `Next.js API Route`.
2. Remove unnecessary configs from `tsconfig` and `next.config`
3. Distinguish the dependencies whether to add on `production` or `development`.
4. Do not add `.env` files to `.git` repository history. Add them to `.gitignore`.

These are the things I could let you know for now and let me know if you need more information about your project.

I have forked your repo and will be going to push a pull request so you can check it out.

![Screenshot 2023-05-31 at 5 24 10 AM](https://github.com/vercel/next.js/assets/120007119/50062ac2-02e4-49ba-b42a-b6c42f797c27)

**Reply by devjiwonchoi** [reply]
Opened a PR: https://github.com/goat6/spacecases-remake/pull/2

**Reply by devvf** [reply]
Thank you very much for the help, it's very appreciated. I will be sure to let you know in the future if anything else arises. Take care :)

**Reply by Coderamrin** [reply]
Thanks [devjiwonchoi](https://github.com/devjiwonchoi) I was having this problem for weeks. 
Thanks for the help 👍

**Reply by devjiwonchoi** [reply]
Glad it was helpful😄 Happy Coding!

**Reply by vuongManabie** [reply]
It's will build successfuly. But Can I ask a question when we change this file to the name "auth.ts" how do we handle sign-in feature?

**Reply by Coderamrin** [reply]
I had the same issue. So I reverted the file name to route.ts. 
then I've added this code snippet. 

`import { getServerSession } from "next-auth/next";
import { NextResponse } from "next/server";
import { authOptions } from "./auth/[...nextauth]/route";

export async function GET(request: Request) {
  const session = await getServerSession(authOptions);

  if (!session) {
    return new NextResponse(JSON.stringify({ error: "unauthorized" }), {
      status: 401,
    });
  }

  return NextResponse.json({ authenticated: !!session });
}`

**Reply by Coderamrin** [reply]
it's on the route.ts file under the api folder

**Reply by vuongManabie** [reply]
It's work . Thank you very much !

**Reply by devvf** [reply, marked as answer]
Hello! I'd just liked to say what worked for me in the end is removing the export in front of `export const authOptions: NextAuthOptions` as that for some reason does not seem to work when building to production - this makes the solution viable for NextJS 13 using route handlers which requires route.ts to be the file for an endpoint. (This applies to my original code). 

I'd just like to say if this doesn't work for you - please refer to @devjiwonchoi's code which can be found above and is also very helpful for a number of optimisations in my project and Next in general.

```
import NextAuth, { NextAuthOptions } from "next-auth";
import DiscordProvider from "next-auth/providers/discord";

const authOptions:NextAuthOptions = {
  providers: [
    DiscordProvider({
      clientId: process.env.DISCORD_CLIENT_ID!,
      clientSecret: process.env.DISCORD_CLIENT_SECRET!,
    })
  ]
};

const handler = NextAuth(authOptions);
export { handler as GET, handler as POST };
```

Good luck with your projects guys 👍  Happy Coding! 

-Thomas

**Reply by Kofichamp** [reply]
Wow pal thanks a lot. Its worked for me

**Reply by AuthorOfTheSurf** [reply]
For my case, I still needed `authOptions: NextAuthOptions` to be exported.

NextJS is only be this strict if the file is named `route.ts`.

So I simply the declaration out of `route.ts`, and into a new file that I called `options.ts` that is in the same `[...nextauth]/` directory

**Reply by HickAndrade** [reply]
route.ts to options.ts?

**Reply by BALDEJ1** [reply]
// options.ts
```
import prisma from '@/lib/prismadb'
import { PrismaAdapter } from '@auth/prisma-adapter'
import { AuthOptions } from 'next-auth'
import GoogleProvider from 'next-auth/providers/google'

const authOptions: AuthOptions = {
  adapter: PrismaAdapter(prisma),
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID as string,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET as string,
    }),
  ],
  secret: process.env.NEXTAUTH_SECRET,
}

export default authOptions;

```

// route.ts
```
import authOptions from './options'; // Update the path accordingly
import NextAuth from 'next-auth';

const handler = NextAuth(authOptions);
export { handler as GET, handler as POST }
```

**Reply by halowahyudi** [reply]
work for me.: next 14.1.0

**Reply by kaumac** [reply]
Just wanted to say that this worked perfectly for me, thank you so much!

**Reply by jamesllopes** [reply]
thanks!!!

**Comment by ghjunior** [comment]
Solution: https://github.com/vercel/next.js/issues/50870#issuecomment-1598655396

**Reply by denniztr** [reply]
This solution has been helpfull to me

"next": "14.1.0"

**Reply by nicitaacom** [reply]
[this](https://github.com/vercel/next.js/issues/50870#issuecomment-1598655396) solution works for me

"next":14.0.4

**Comment by absolutegravitas** [comment]
existing documentation here is still not updated because of this -- how are newbies just implementing supposed to know

**Comment by ffan0811** [comment]
I'm getting this error after upgrading nextjs to 14

**Reply by phamvant** [reply]
@ffan0811 have you fixed it yet?

**Reply by burn-c** [reply]
@ffan0811 
Provisionally I created a new file with the authOptions function and imported it into the file `src/app/api/auth/[...nextauth]/route.ts` and in other places I was using authOptions, in my case resolved the type error.

**Reply by Vancelott** [reply]
@burn-c Did you name the file auth.ts, and where did you exactly place it? Thank you in advance.

**Reply by burn-c** [reply]
@Vancelott, since it's a specific context of NextAuth, I placed it within the same folder ( `src/app/api/auth/[...nextauth]/authOptions.ts` ) . As Next.js validates only 'route.ts' files, I believe it's not an issue to leave it there."

**Reply by Vancelott** [reply]
@burn-c Thank you, but I've got it working already as well. I created a new file in the root of the project called auth.ts, where I placed my authOptions, but I think what solved it for me was the import of the correct type for the authOptions - NextAuthOptions, and not just AuthOptions. One more thing I was always forgetting is to import the new authOptoins to my getSession hook.

**Reply by mergd** [reply]
Combo of moving authOptions into it's own file and removing the .ts from route.ts did the trick.

With the route.ts, it kept having the error `"handler" is not a valid Route export field.`

**Comment by whitallee** [comment]
For some reason I had `"export const handler = NextAuth(authOptions);"`. I did attempt first by moving the `authOptions` to a utils file, but it didnt fix it by itself, however removing the `export` before `const handler` is what fully fixed it for me.

**Reply by andreimacavei** [reply]
This was the problem for me as well, I had a leftover export in front of handler

**export** const handler = NextAuth(authOptions);
export { handler as GET, handler as POST };

**Reply by nestorvave** [reply]
thank u

**Comment by nicitaacom** [comment]
[this](https://github.com/vercel/next.js/issues/50870#issuecomment-1598655396) solution works for me

"next":14.0.4

**Comment by brytebee** [comment]
Thanks!

The solution worked! 👍

**Comment by stellaoliveirabertt** [comment]
### ✅ Definitive Solution for Next.js 15 with `next-auth@5`

If you're getting this build error:

```bash
Type error: Route "src/app/api/auth/[...nextauth]/route.js" does not match the required types of a Next.js Route.
"authOptions" is not a valid Route export field.
```

⚠️ It means that **Next.js 15 no longer allows exporting helper objects (like `authOptions`) directly from route files** such as `app/api/**/route.js` or `.ts`.

---

#### ✅ How to fix it properly:

---

### 1. Create a separate file for your `authOptions`

**Suggested path:** `src/app/lib/auth.js`

```js
// src/app/lib/auth.js
import AzureADProvider from "next-auth/providers/azure-ad";

const authOptions = {
  providers: [
    AzureADProvider({
      clientId: process.env.AZURE_AD_CLIENT_ID,
      clientSecret: process.env.AZURE_AD_CLIENT_SECRET,
      tenantId: process.env.AZURE_AD_TENANT_ID,
    }),
  ],
  pages: {
    signIn: "/dashboard/login",
  },
  callbacks: {
    async session({ session, token }) {
      session.user.id = token.sub;
      return session;
    },
  },
};

export default authOptions;
```

---

### 2. Then fix your `[...nextauth]/route.js` (or `.ts`) like this:

```js
import NextAuth from "next-auth";
import authOptions from "@/app/lib/auth";

const handler = NextAuth(authOptions);

export const GET = handler;
export const POST = handler;
```

---

### ✅ Now run:

```bash
npm run build
```

Your app should now build successfully on **Next.js 15 + `next-auth` v5**.
